compiler/src/dotty/tools/dotc/core/GadtConstraint.scala

```diff
@@ -71,12 +71,26 @@ class GadtConstraint private (
     externalize(constraint.nonParamBounds(param)).bounds
 
   def fullLowerBound(param: TypeParamRef)(using Context): Type =
-    constraint.minLower(param).foldLeft(nonParamBounds(param).lo) {
+    val self = externalize(param)
+    constraint.minLower(param).filterNot { p =>
+      val sym = paramSymbol(p)
+      sym.name.is(NameKinds.UniqueName) && {
+        val hi = sym.info.hiBound
+        !hi.isExactlyAny && self <:< hi
```

Comment:
Would be good to add a comment explaining why this is needed and pointing to a test case.